src/main/java/com/google/jenkins/plugins/credentials/oauth/JsonServiceAccountConfig.java

```diff
@@ -53,36 +63,109 @@
   private static final long serialVersionUID = 6818111194672325387L;
   private static final Logger LOGGER =
       Logger.getLogger(JsonServiceAccountConfig.class.getSimpleName());
-  private String jsonKeyFile;
+  @CheckForNull
+  private String filename;
+  @CheckForNull
+  private SecretBytes secretJsonKey;
+  @Deprecated   // for migration purpose
+  @CheckForNull
+  private transient String jsonKeyFile;
   private transient JsonKey jsonKey;
 
+  /** @since 0.8 */
   @DataBoundConstructor
-  public JsonServiceAccountConfig(FileItem jsonKeyFile,
-      String prevJsonKeyFile) {
-    if (jsonKeyFile != null && jsonKeyFile.getSize() > 0) {
+  public JsonServiceAccountConfig() {}
+
+  /**
+   * For being able to load credentials created with versions < 0.8
+   * and backwards compatibility with external callers.
+   *
+   * @param jsonKeyFile The uploaded json key file
+   * @param prevJsonKeyFile The path of the previous json key file
+   * @since 0.3
+   */
+  @Deprecated
+  public JsonServiceAccountConfig(
+      FileItem jsonKeyFile, String prevJsonKeyFile) {
+    this.setJsonKeyFileUpload(jsonKeyFile);
+    if (filename == null && prevJsonKeyFile != null) {
+      this.filename = extractFilename(prevJsonKeyFile);
+      this.secretJsonKey = getSecretBytesFromFile(prevJsonKeyFile);
+    }
```

Comment:
What is the behavior if filename and secretJsonKey are still null at this point? Is it worth logging this case and/or throwing an exception for this case?

Reply:
If both are still null, then they would first see a log message from the call to setUploadJsonKeyFile. With both null, the credentials created wouldn't be usable so I think at the very least adding a log message here makes sense. I've purposefully avoided using exceptions in the databound setters to avoid the demon jenkins page, but for this constructor it may make sense to throw an exception instead because it is not called directly in the UI.